Node level bucket heal

shtripat opened the pull request:
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description


## Motivation and Context
Re-enabling bucket heal at node level with default setting of `SetCount` in `HealResultItem` response to client.

## How to test this PR?


## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

shtripat committed:
Heal buckets at node level

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Incorporated review comments

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Use objAPI directly for healing bucket while global heal

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Populate SetCount in HealResultItem

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

vadmeste commented on cmd/erasure-server-pool.go:
```diff
@@ -1941,8 +1941,55 @@ func (z *erasureServerPools) HealBucket(ctx context.Context, bucket string, opts
+				if disk == OfflineDisk {
+					continue
+				}
+				vi, err := disk.StatVol(ctx, bucket)
```
it doesn't seem we are checking for this error here

shtripat replied:
This `err` would be added to the disk errors list down. Based on below comments action, this would be added even for first disk of first erasure set.

vadmeste commented on cmd/erasure-server-pool.go:
```diff
@@ -1941,8 +1941,55 @@ func (z *erasureServerPools) HealBucket(ctx context.Context, bucket string, opts
+				ds, ok := diskStats[hostName]
+				if !ok {
+					newds := DiskStat{
+						VolInfos: vis,
```
I am not able to understand this logic; 'vis' can be empty here for the first disk in the first erasure set.

shtripat replied:
Ah, that's good catch. Will add the volume info and err to the lists and then set as `newds`.

vadmeste commented on cmd/erasure-server-pool.go:
```diff
@@ -1941,8 +1941,55 @@ func (z *erasureServerPools) HealBucket(ctx context.Context, bucket string, opts
+				bktNotFoundCount++
+			}
+		}
+		// if the bucket if not found on more than hslf the no of nodes, its dagling
```
s/dagling/dangling

vadmeste reviewed (commented):
Few comments, still need another review round

shtripat committed:
More review comments

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Wait longer after `minio server`

With recent changes it takes little more time to mark the MinIO
instance ready.

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

vadmeste commented on cmd/peer-s3-server.go:
```diff
@@ -78,6 +81,134 @@ func (s *peerS3Server) HealthHandler(w http.ResponseWriter, r *http.Request) {
+				afterState[index] = madmin.DriveStateMissing
+
+				// mutate only if not a dry-run
+				if opts.DryRun {
```
I don't think this is the right place to return after a dry-run here because this is a separate goroutine and I didn't see how the code will stop executing DeleteVol later

shtripat replied:
This function was copied as is from erasure-healing.go. We may need to have check for dry run condition outside for not mutating the bucket?

shtripat committed:
Dont mutate if dry run (at right place)

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Merge branch 'master' into node-level-bucket-heal

shtripat committed:
Merge branch 'master' into node-level-bucket-heal

vadmeste reviewed (approved):
LGTM
